src/nomad_simulations/schema_packages/basis_set.py

```diff
@@ -0,0 +1,642 @@
+from nomad.datamodel.data import ArchiveSection
+from nomad.datamodel.datamodel import EntryArchive
+from nomad.datamodel.metainfo.annotations import ELNAnnotation
+from nomad.metainfo import MEnum, Quantity, SubSection
+from nomad.units import ureg
+import itertools
+import numpy as np
+import pint
+from scipy import constants as const
+from structlog.stdlib import BoundLogger
+from typing import Optional, Any, Callable
+
+from nomad_simulations.schema_packages.atoms_state import AtomsState
+from nomad_simulations.schema_packages.numerical_settings import (
+    Mesh,
+    NumericalSettings,
+)
+from nomad_simulations.schema_packages.properties.energies import EnergyContribution
+
+
+def check_normalized(func: Callable):
+    """
+    Decorator to check if the section is already normalized.
+    """
+
+    def wrapper(self, archive: EntryArchive, logger: BoundLogger) -> None:
+        if self._is_normalized:
+            return None
+        func(self, archive, logger)
+        self._is_normalized = True
+
+    return wrapper
+
+
+class BasisSet(NumericalSettings):
+    """A type section denoting a basis set component of a simulation.
+    Should be used as a base section for more specialized sections.
+    Allows for denoting the basis set's _scope_, i.e. to which entity it applies,
+    e.g. atoms species, orbital type, Hamiltonian term.
+
+    Examples include:
+    - mesh-based basis sets, e.g. (projector-)(augmented) plane-wave basis sets
+    - atom-centered basis sets, e.g. Gaussian-type basis sets, Slater-type orbitals, muffin-tin orbitals
+    """
+
+    name = Quantity(
+        type=str,
+        description="""
+        Name of the basis set component.
+        """,
+    )
+
+    species_scope = Quantity(
+        type=AtomsState,
+        shape=['*'],
+        description="""
+        Reference to the section `AtomsState` specifying the localization of the basis set.
+        """,
+        a_eln=ELNAnnotation(components='ReferenceEditQuantity'),
+    )
+
+    # TODO: add atom index-based instantiator for species if not present
+
+    hamiltonian_scope = Quantity(
+        type=EnergyContribution,
+        shape=['*'],
+        description="""
+        Reference to the section `EnergyContribution` containing the information
+        of the Hamiltonian term to which the basis set applies.
+        """,
+        a_eln=ELNAnnotation(components='ReferenceEditQuantity'),
+    )
+
+    # ? band_scope or orbital_scope: valence vs core
+
+    def normalize(self, archive, logger):
+        super().normalize(archive, logger)
+        self.name = self.m_def.name
```

Comment:
Is this assigning `BasisSet` to all the other basis set classes? I think you should write the `normalize` function on each of the basis set so it assigns properly `'PlaneWaveBasisSet'`, etc.

Reply:
Nope, `self` refers to the calling class.